### PR TITLE
WIP: impl `derp::client` mod

### PR DESCRIPTION
**Pull request by ramfox**
some general notes about this first attempt (aka "this mess"):

- golang impl uses u32 everywhere, including sending the frame length on the wire.  I assuming we should continue to only use 4 bytes for the frame length, but then we should probably convert to `usize` when we get it, so that we have fewer `as usize`'s littered throughout the code. But I haven't done this yet.

- from looking at the surrounding code that imports the derp client, it looks like we need to be able to use this in many threads. So I've wrapped any "shared data" in an `Arc<Mutex>`

- I don't understand why the go code only wrapped the `writer` in a mutex lock and not the `reader` as well. I followed the same pattern, because maybe I don't understand this yet?

- the rate limiter is gross and maybe we wrap this in a trait so we don't have to drag around the underlying `governor` traits in our client implementation. Updating the rate limiter is also not fully implemented. Because I'm mad at how gross it is & also I needed to stop spending time on it & move on.

- this `recv_err` nonsense is so ugly, god help me. After we encounter an error reading while running `recv`, the go impl wants us to return the same error anytime we call `recv` again on the client. (The client is "dead" after the first error). However, most errors in rust aren't clonable, so currently I'm storing the error as a `String` that I've pulled from the error itself. There has got to be a better way to do this, other than the nonsense I've implemented, no?

### CODE REVIEW

File: src/hp/derp/client.rs [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+                }
+                read_total += read;
+            }
+            if read_total != frame_len as usize {
+                bail!(
+                    "unexpected number of bytes sent in frame, said {}, received {}",
+                    frame_len,
+                    read_total
+                );
+            }
```
Comment (flub) [inline src/hp/derp/client.rs:338]: aren't you re-implementing `read_exact()`?  if you create the frame_payload after you know frame_len you might be able to use that instead?

File: src/hp/derp/client.rs [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+    pub async fn recv(&self) -> Result<ReceivedMessage> {
+        self.recv_check_error(Duration::from_secs(120)).await
+    }
+
+    async fn recv_check_error(&self, timeout: Duration) -> Result<ReceivedMessage> {
+        {
+            let recv_err = self.recv_err.lock().await;
+            if let Some(err) = &*recv_err {
+                bail!(err.clone());
+            }
```
Comment (flub) [inline src/hp/derp/client.rs:295]: do you really care about the exact error?  We use anyhow errors anyway so i don't think we do.  you can just store an (atomic) bool for the error state for now i think.

File: src/hp/derp/client.rs [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+        {
+            bail!("invalid server greeting");
+        }
+        let key: [u8; 32] = buf[magic_len..magic_len + 32].try_into()?;
+        self.server_key = key::node::PublicKey::from(key);
+        Ok(())
+    }
+
+    async fn parse_server_info(&self, buf: &mut [u8]) -> Result<ServerInfo> {
+        // TODO: should NONCE_LEN just be usize?
```
[the remark this answers is not in the document]
Comment (ramfox) [reply src/hp/derp/client.rs]: changed the constants to `usize`

File: src/hp/derp/client.rs [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+        if packet.len() > MAX_PACKET_SIZE {
+            bail!("packet too big: {}", packet.len());
+        }
+        let frame_len = u32::try_from(key::node::KEY_SIZE + packet.len())?;
+        let rate_limiter = match &self.rate_limiter {
+            None => None,
+            Some(rl) => Some(Arc::clone(&rl)),
+        };
+        {
+            let mut writer = self.writer.lock().await;
```
[the remark this answers is not in the document]
Comment (ramfox) [reply src/hp/derp/client.rs:153]: okay thank you!

File: src/hp/derp/client.rs (hunk 2 above)
Comment (ramfox) [reply src/hp/derp/client.rs:294]: I... don't know if we care about the exact error. Maybe for now, we go the atomic bool route & just return an anyhow error saying something like "client has encountered a read error and can no longer receive messages"

File: src/hp/derp/client.rs (hunk 1 above)
Comment (ramfox) [reply src/hp/derp/client.rs]: yes, I did... I basically couldn't figure out how to specify the exact length of the buffer I wanted to use... until just now realizing I could pass a slice of a specific size.

Thanks!

File: src/hp/derp/client.rs [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+impl<W, R, S, C, MW> Client<W, R, S, C, MW>
+where
+    W: AsyncWrite + Send + Unpin + 'static,
+    R: AsyncRead + Unpin,
+    S: governor::state::DirectStateStore<Key = governor::state::direct::NotKeyed>,
+    C: governor::clock::Clock,
+    MW: governor::middleware::RateLimitingMiddleware<C::Instant>,
+{
+    // TODO: for something relatively straight forward, this is pretty hard to follow
+    // TODO: also, should Client.server_key be an option?
```
[the remark this answers is not in the document]
Comment (ramfox) [reply src/hp/derp/client.rs]: I'm just going to leave this for now because I think part of the problem is the original design for `read_frame` that the go impl has, and all the reading / writing needs to be looked at more carefully anyway. But thank you for dropping some knowledge here.

**Comment by ramfox** [comment]
Okay more notes:
- `const`s are now all `usize`, except the `FrameType`s
- removed all `Mutex`s (including the one on the writer) until rust yells at us
- wrote a `RateLimiter` struct that hides all the ugliness of hanging onto the `governor::RateLimiter`
- cleaned up some `PublicKey` vs `SecretKey` vs `SharedKey` nonsense
- using `AtomicBool` to track if the client is dead

things I'm worried about:
- there are too many ways we attempt to read and write things & I don't think I implemented the most efficient way to do any of them. I hope the surface area of this gets smaller and more correct as we refactor
- we need to talk about the `mesh_key`. In the go impl it's a string. I make it an array because we send the `mesh_key` over the wire & I wanted to use postcard to serialize the struct (and postcard can't estimate the `POSTCARD_MAX_SIZE` with `String` fields.) It was just the quickest way for me to get this done, but we need to figure out the "right" way at some point.

**Review by flub — APPROVED** [review]
there's a couple of blocking bugs wrt to reading from the streams, but i don't want to hold this up once they're fixed

File: src/hp/derp.rs [hunk 6]
```diff
@@ -163,23 +163,25 @@ async fn read_frame_header(mut reader: impl AsyncRead + Unpin) -> Result<(FrameT
...
+    max_size: usize,
+    mut bytes: &mut BytesMut,
+) -> Result<(FrameType, usize)> {
     let (frame_type, frame_len) = read_frame_header(&mut reader).await?;
     if frame_len > max_size {
         bail!("frame header size {frame_len} exceeds reader limit of {max_size}");
     }
-
+    // TODO: trusting that the caller has sized the buffer correctly? Shouldn't we just attempt to
+    // read `frame_len` amount?
```
[the remark this answers is not in the document]
Comment (ramfox) [reply src/hp/derp.rs:174]: 🔥 thank you

File: src/hp/derp/client.rs [hunk 7]
```diff
@@ -1,21 +1,532 @@
...
+        Ok(())
+    }
+
+    pub async fn forward_packet(
+        &mut self,
+        srckey: PublicKey,
+        dstkey: PublicKey,
+        packet: &[u8],
+    ) -> Result<()> {
+        tokio::select! {
```
[the remark this answers is not in the document]
Comment (ramfox) [reply src/hp/derp/client.rs:148]: ahh I thought this only really mattered in loops. adding it in.

File: src/hp/derp/client.rs [hunk 8]
```diff
@@ -1,21 +1,532 @@
...
+                res
+            }
+        }
+    }
+
+    async fn recv_0(&mut self) -> Result<ReceivedMessage> {
+        // from the tailscale/derp/derp_client.go:
+        // "In practice [a frame is] 4KB (from derphttp.Client's bufio.NewReader(httpConn)) and
+        // in practice, WireGuard packets (and thus DERP frames) are under 1.5 KiB."
+        let mut frame_payload = BytesMut::with_capacity(4 * 1024);
```
[the remark this answers is not in the document]
Comment (ramfox) [reply src/hp/derp/client.rs:291]: Ah, no they don't allocate the 4KB explicitly. I just read this backwards.

We aren't using WireGuard frames tho, we are using quic frames (over HTTP) right? Those seem to be less than 1.5KB as well (https://blog.apnic.net/2019/03/04/a-quick-look-at-quic/#:~:text=QUIC%20and%20IP%20fragmentation&text=For%20IPv4%2C%20the%20QUIC%20maximum,is%201%2C392%20packets%20in%20size), so I'll change this.

File: src/hp/derp/client.rs [hunk 9]
```diff
@@ -1,21 +1,532 @@
...
+        loop {
+            let (frame_type, frame_len) = read_frame_header(&mut self.reader).await?;
+            if frame_len > MAX_FRAME_SIZE {
+                bail!("unexpectedly large frame of {} bytes returned", frame_len);
+            }
+            self.reader
+                .read_exact(&mut frame_payload[..frame_len])
```
[the remark this answers is not in the document]
Comment (ramfox) [reply src/hp/derp/client.rs:298]: thank you

**Comment by ramfox** [comment]
fixed up the reading and writing story & added read/write round trip tests

File: src/hp/derp/client.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (flub) [reply src/hp/derp/client.rs]: QUIC frames could be anything above 1200 bytes.  But QUIC is supposed to find the path MTU so they never get fragmented packets and realistically on the internet this is unlikely to exceed 1500 as that's the standard ethernet frame size.
